pkg/s3signer/request-signature-streaming.go

```diff
@@ -0,0 +1,242 @@
+
+	chunkStringToSign := getChunkStringToSign(reqTime, location, previousSignature, chunkData)
+	signingKey := getSigningKey(secretAccessKey, location, reqTime)
+	return getSignature(signingKey, chunkStringToSign)
+}
+
+// getSeedSignature - returns the seed signature for a given request.
+func (s *StreamingReader) setSeedSignature(req *http.Request) {
+	// Get canonical request
+	canonicalRequest := getCanonicalRequest(*req, ignoredStreamingHeaders)
+
+	// Get string to sign from canonical request.
+	stringToSign := getStringToSignV4(s.reqTime, s.location, canonicalRequest)
+
+	signingKey := getSigningKey(s.secretAccessKey, s.location, s.reqTime)
+
+	// Calculate signature.
+	s.seedSignature = getSignature(signingKey, stringToSign)
+}
+
+// Streaming signature as a reader on top of req.Body's ReaderCloser
+// chunk header;data;... repeat
+type StreamingReader struct {
+	accessKeyID     string
+	secretAccessKey string
+	location        string
+	prevSignature   string
+	seedSignature   string
+	contentLen      int64
+	baseReadCloser  io.ReadCloser
+	bytesRead       int64
+	buf             bytes.Buffer
+	chunkBuf        []byte
+	done            bool
+	reqTime         time.Time
+}
+
+// SetStreamingAuthHeader - builds and sets authorization header value for streaming signature.
+func (s *StreamingReader) SetStreamingAuthHeader(req *http.Request) {
+	credential := GetCredential(s.accessKeyID, s.location, s.reqTime)
+	authParts := []string{
+		signV4Algorithm + " Credential=" + credential,
+		"SignedHeaders=" + getSignedHeaders(*req, ignoredStreamingHeaders),
+		"Signature=" + s.seedSignature,
+	}
+
+	// Set authorization header.
+	auth := strings.Join(authParts, ",")
+	req.Header.Set("Authorization", auth)
+}
+
+// NewStreamingReader - provides chunked upload signatureV4 support by
+// implementing io.Reader.
+func NewStreamingReader(req *http.Request, reader io.ReadCloser, accessKeyID, secretAccessKey, location string, reqTime time.Time) (*StreamingReader, error) {
+	stReader := &StreamingReader{
+		baseReadCloser:  reader,
+		accessKeyID:     accessKeyID,
+		secretAccessKey: secretAccessKey,
+		location:        location,
+		reqTime:         reqTime,
+		chunkBuf:        make([]byte, payloadChunkSize),
+	}
+	// Get Content-Length from request header.
+	dataLen, err := strconv.ParseInt(req.Header.Get("Content-Length"), 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	stReader.contentLen = dataLen
+	stReader.setSeedSignature(req)
+	// Set seed signature as prevSignature for subsequent
+	// streaming signing process.
+	stReader.prevSignature = stReader.seedSignature
+	return stReader, nil
+}
+
+func (s *StreamingReader) Read(buf []byte) (int, error) {
+	if s.done {
+		// return what is left from last read from baseReader.
+		// s.buf will return io.EOF once the buffer is drained.
+		return s.buf.Read(buf)
+	}
+
+	n0, err := s.buf.Read(buf)
+	if err == io.EOF {
```

Comment:
s.buf.Read() is called in different places in this function. I think this can be simplified this way: the type of s.buf is bytes.Buffer so we can know the length of data inside. What we need is to fill s.buf whenever len(s.buf) < len(buf) and then stream the content of s.buf at the end of this function, do you think this doable @krisis ?